--- schroedinger\schroasync-win32.c	Sun Mar 22 14:22:33 2009
+++ schroedinger\schroasync-win32.c	Mon Mar 30 09:59:27 2009
@@ -166,8 +166,8 @@
 schro_async_start (SchroAsync *async)
 {
   int i;
-  for(i=0;i<async->n_threads) {
-    SetEvent (async->thread_event);
+  for(i=0;i<async->n_threads;i++) {
+    SetEvent (async->threads[i].event);
   }
 }
 
@@ -332,7 +332,7 @@
      * a lot of unnecessary wakeups in some cases. */
     {
       int i;
-      for(i=0;i<async->n_threads) {
+      for(i=0;i<async->n_threads;i++) {
         SetEvent (async->thread_event);
       }
     }
